ocaml/lambda/matching.ml

```diff
@@ -3424,14 +3352,19 @@ let rec compile_match ~scopes value_kind repr partial ctx
   match m.cases with
   | ([], action) :: rem ->
       (match action with
-        | Unguarded action ->
-            event_branch repr action, Jumps.empty
-        | Guarded { patch_guarded; _ } ->
-            let lambda, total =
-              compile_match ~scopes value_kind None partial ctx
-                { m with cases = rem }
-            in
-            event_branch repr (patch_guarded ~patch:lambda), total)
+       | Unguarded action -> event_branch repr action, Jumps.empty
+       | Guarded { action; exn_label } ->
+           let lambda, total =
+             compile_match ~scopes value_kind None partial ctx
+               { m with cases = rem }
+           in
+           let rec catch_fallthrough = function
+             | Llet (str, k, id, lam, body) ->
+                 Llet (str, k, id, lam, catch_fallthrough body)
+             | Levent (lam, ev) -> Levent (catch_fallthrough lam, ev)
+             | lam ->  Lstaticcatch (lam, (exn_label, []), lambda, value_kind)
```

Comment:
Why is this necessary? Naively, I'd have guessed that inserting `Lstaticcatch` around the whole expression is always ok.

Reply:
The result of this operation is passed to `event_branch`, which gets mad when `repr` is not `None` and `lam` is not some series of `Llet`s pointing to an `Levent` (as in this case, `repr` is meant to be stitched into the `Levent`).

I think the better way to handle this is to change `event_branch` to traverse the bodies of `Lstaticcatch`s when looking for the appropriate `Levent`, which is what I've done.